Api end points question controller - QuestionEntity (Completed), QuestionDAO (Completed) , QuestionController (WIP)

siddhartth1987 opened the pull request:
Implementation of QuestionEntity and QuestionDAO classes is complete. 
QuestionController is still a work in progress.

siddhartth1987 committed:
Serializable Entity Class that corresponds to the 'question' table.

siddhartth1987 committed:
API endpoint implementations within the Question Controller class (REST API endpoints - 2); Added skeletons for createQuestion, deleteQuestion, editQuestionContent

siddhartth1987 committed:
Merge remote-tracking branch 'origin/master' into APIEndPointsQuestionController

siddhartth1987 committed:
DAO class for accessing information in question table

Sanjit0891 commented:
Seems like the request is missing few stuff like Service and User auth hence will not be merging it atm.

siddhartth1987 committed:
Added QuestionBusinessService that'll liase with  QuestionDAO to perform operations based on requests from QuestionController

Added calls to QuestionBusinessService from QuestionController

siddhartth1987 committed:
Merge remote-tracking branch 'origin/master' into APIEndPointsQuestionController

# Conflicts:
#	quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java

siddhartth1987 committed:
Made a small change to method signature that is calling UserDAO to get user details using userID

siddhartth1987 committed:
AuthenticationService : Added validateAuthToken method to validate accessToken of inbound request

RestExceptionHandler : Added userNotFoundException and AuthorizationFailedException

siddhartth1987 committed:
Made a slight modification to getUser method; Changed Set parameter to uuid from userUuid

Sanjit0891 commented on quora-api/src/main/java/com/upgrad/quora/api/controller/QuestionController.java:
```diff
@@ -0,0 +1,98 @@
+
+    // This method will create a new question posted by an authorized user.
+    @RequestMapping(method = RequestMethod.POST, path = "/question/create", consumes = MediaType.APPLICATION_JSON_UTF8_VALUE, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<QuestionResponse> createQuestion(
```
i think you need to have user auth service to get the token and authenticate - 

This endpoint requests for all the attributes in 'QuestionRequest' about the question and access token of the signed in user as a string in the authorization field of the Request Header.

siddhartth1987 replied:
Why is auth service required in this class? That's being taken care of the in the Question Business Service class.

siddhartth1987 replied:
The controller will pass authorization parameter to the business service which will, in turn, validate the authorization before performing any operations.

Sanjit0891 commented on quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java:
```diff
@@ -18,10 +20,29 @@
-    @ExceptionHandler(AuthenticationFailedException.class)
-    public ResponseEntity<ErrorResponse> AuthenticationFailedException(AuthenticationFailedException exe, WebRequest request) {
+    @ExceptionHandler(AuthorizationFailedException.class)
+    public ResponseEntity<ErrorResponse> authorizationFailedException(
```
You should create a new authfailexception rather than amending the existing one as it will break the signin controller

siddhartth1987 replied:
I didn't realize that I'd overridden this .. Will make the change.

Sanjit0891 commented on quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java:
```diff
@@ -21,7 +21,7 @@ public UserEntity createUser(UserEntity userEntity) {
     public UserEntity getUser(final String userUuid) {
         try {
-            return entityManager.createNamedQuery("userByUuid", UserEntity.class).setParameter("userUuid", userUuid).getSingleResult();
+            return entityManager.createNamedQuery("userByUuid", UserEntity.class).setParameter("uuid", userUuid).getSingleResult();
```
sorry i mis read it but i think it is correct to have userUuid as it is being used this way in JWT

siddhartth1987 replied:
Please look at the parameter defined in the NamedQuery "userByUuid"  in UserEntity
@NamedQuery(name = "userByUuid", query = "select u from UserEntity u where u.uuid =:uuid")

siddhartth1987 replied:
The above change is in accordance with what has been specified in UserEntity

siddhartth1987 committed:
Merged AuthenticationFailedException from master branch and added  AuthorizationFailedException, InvalidQuestionException, UserNotFoundException

siddhartth1987 commented:
Changes have either been made in accordance with review comments or appropriate responses have been added. Merging with master branch.
